File: figma-clipboard/kiwi.py
"""Minimal Kiwi schema + data decoder.

Format reference: https://github.com/evanw/kiwi/blob/master/binary.md

Wire format primer:
  * varint (unsigned): LEB128 little-endian, 7 bits per byte, MSB = continuation
  * int (signed):      zigzag varint
  * string:            null-terminated UTF-8

Schema layout:
  varint definition_count
  for each definition:
    string name
    uint8  kind     (0 = enum, 1 = struct, 2 = message)
    varint field_count
    for each field:
      string name
      int    type   (negative = builtin code, 0-based index into definitions otherwise)
      uint8  is_array
      varint value  (enum constant, struct = 0, message = field id)

Builtin type codes:
  -1 bool   -2 byte   -3 int    -4 uint
  -5 float  -6 string -7 int64  -8 uint64

Data layout (recursive per definition):
  - enum:    varint  → field_index lookup → field.value (a constant)
  - struct:  fields in declaration order, no terminator
  - message: pairs of (varint field_id, value) terminated by field_id = 0
  - array:   varint count, then count values
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from io import BytesIO
from typing import Any


BUILTINS = {
    -1: "bool",
    -2: "byte",
    -3: "int",
    -4: "uint",
    -5: "float",
    -6: "string",
    -7: "int64",
    -8: "uint64",
}

KIND_ENUM = 0
KIND_STRUCT = 1
KIND_MESSAGE = 2
KIND_NAMES = {0: "enum", 1: "struct", 2: "message"}


@dataclass
class FieldDef:
    name: str
    type_code: int
    is_array: bool
    value: int


@dataclass
class Definition:
    name: str
    kind: int
    fields: list[FieldDef] = field(default_factory=list)


class ByteReader:
    __slots__ = ("buf", "pos")

    def __init__(self, buf: bytes, pos: int = 0) -> None:
        self.buf = buf
        self.pos = pos

    def remaining(self) -> int:
        return len(self.buf) - self.pos

    def read_byte(self) -> int:
        b = self.buf[self.pos]
        self.pos += 1
        return b

    def read_bytes(self, n: int) -> bytes:
        out = self.buf[self.pos : self.pos + n]
        self.pos += n
        return out

    def read_varint(self) -> int:
        result = 0
        shift = 0
        while True:
            b = self.buf[self.pos]
            self.pos += 1
            result |= (b & 0x7F) << shift
            if (b & 0x80) == 0:
                return result
            shift += 7
            if shift > 63:
                raise ValueError("varint too long")

    def read_signed_varint(self) -> int:
        v = self.read_varint()
        return (v >> 1) ^ -(v & 1)

    def read_bool(self) -> bool:
        return self.read_byte() != 0

    def read_int(self) -> int:
        return self.read_signed_varint()

    def read_uint(self) -> int:
        return self.read_varint()

    def read_int64(self) -> int:
        return self.read_signed_varint()

    def read_uint64(self) -> int:
        return self.read_varint()

    def read_float(self) -> float:
        first = self.read_byte()
        if first == 0:
            return 0.0
        rest = self.read_bytes(3)
        bits = first | (rest[0] << 8) | (rest[1] << 16) | (rest[2] << 24)
        bits = ((bits & 0xFFFFFFFF) << 23) | ((bits & 0xFFFFFFFF) >> 9)
        bits &= 0xFFFFFFFF
        return struct.unpack("<f", struct.pack("<I", bits))[0]

    def read_string(self) -> str:
        start = self.pos
        end = self.buf.index(0, start)
        s = self.buf[start:end].decode("utf-8")
        self.pos = end + 1
        return s


class ByteWriter:
    __slots__ = ("out",)

    def __init__(self) -> None:
        self.out = bytearray()

    def write_byte(self, b: int) -> None:
        self.out.append(b & 0xFF)

    def write_bytes(self, b: bytes) -> None:
        self.out.extend(b)

    def write_varint(self, v: int) -> None:
        if v < 0:
            raise ValueError("varint must be non-negative")
        while True:
            b = v & 0x7F
            v >>= 7
            if v:
                self.out.append(b | 0x80)
            else:
                self.out.append(b)
                return

    def write_signed_varint(self, v: int) -> None:
        zig = ((-v) << 1) - 1 if v < 0 else v << 1
        self.write_varint(zig)

    def write_bool(self, v: bool) -> None:
        self.write_byte(1 if v else 0)

    def write_int(self, v: int) -> None:
        self.write_signed_varint(v)

    def write_uint(self, v: int) -> None:
        self.write_varint(v)

    def write_int64(self, v: int) -> None:
        self.write_signed_varint(v)

    def write_uint64(self, v: int) -> None:
        self.write_varint(v)

    def write_float(self, v: float) -> None:
        if v == 0:
            self.write_byte(0)
            return
        bits = struct.unpack("<I", struct.pack("<f", v))[0]
        rotated = ((bits << 9) | (bits >> 23)) & 0xFFFFFFFF
        # 4 bytes LE
        self.out.append(rotated & 0xFF)
        self.out.append((rotated >> 8) & 0xFF)
        self.out.append((rotated >> 16) & 0xFF)
        self.out.append((rotated >> 24) & 0xFF)

    def write_string(self, s: str) -> None:
        self.out.extend(s.encode("utf-8"))
        self.out.append(0)

    def get(self) -> bytes:
        return bytes(self.out)


# ---------- Schema ----------


def parse_schema(buf: bytes) -> list[Definition]:
    r = ByteReader(buf)
    count = r.read_varint()
    defs: list[Definition] = []
    for _ in range(count):
        name = r.read_string()
        kind = r.read_byte()
        field_count = r.read_varint()
        fields: list[FieldDef] = []
        for _ in range(field_count):
            fname = r.read_string()
            ftype = r.read_signed_varint()
            farr = r.read_byte()
            fval = r.read_varint()
            fields.append(FieldDef(fname, ftype, farr == 1, fval))
        defs.append(Definition(name, kind, fields))
    return defs


def schema_to_dict(defs: list[Definition]) -> dict[str, Any]:
    out = {"count": len(defs), "definitions": []}
    for d in defs:
        entry = {"name": d.name, "kind": KIND_NAMES[d.kind], "fields": []}
        for f in d.fields:
            ref = None
            if f.type_code < 0:
                ref = BUILTINS.get(f.type_code, f"<builtin {f.type_code}>")
            else:
                idx = f.type_code
                if d.kind == KIND_ENUM:
                    ref = None
                elif 0 <= idx < len(defs):
                    ref = defs[idx].name
                else:
                    ref = f"<def {f.type_code}>"
            entry["fields"].append(
                {
                    "name": f.name,
                    "type": ref,
                    "array": f.is_array,
                    "value": f.value,
                }
            )
        out["definitions"].append(entry)
    return out


# ---------- Data ----------


def decode_value(
    reader: ByteReader, type_code: int, is_array: bool, defs: list[Definition]
) -> Any:
    if is_array:
        n = reader.read_varint()
        return [decode_value(reader, type_code, False, defs) for _ in range(n)]
    if type_code < 0:
        return _decode_builtin(reader, type_code)
    return decode_definition(reader, type_code, defs)


def _decode_builtin(reader: ByteReader, code: int) -> Any:
    if code == -1:
        return reader.read_bool()
    if code == -2:
        return reader.read_byte()
    if code == -3:
        return reader.read_int()
    if code == -4:
        return reader.read_uint()
    if code == -5:
        return reader.read_float()
    if code == -6:
        return reader.read_string()
    if code == -7:
        return reader.read_int64()
    if code == -8:
        return reader.read_uint64()
    raise ValueError(f"unknown builtin {code}")


def encode_value(
    writer: ByteWriter,
    value: Any,
    type_code: int,
    is_array: bool,
    defs: list[Definition],
) -> None:
    if is_array:
        writer.write_varint(len(value))
        for item in value:
            encode_value(writer, item, type_code, False, defs)
        return
    if type_code < 0:
        _encode_builtin(writer, value, type_code)
        return
    encode_definition(writer, value, type_code, defs)


def _encode_builtin(writer: ByteWriter, value: Any, code: int) -> None:
    if code == -1:
        writer.write_bool(bool(value))
    elif code == -2:
        writer.write_byte(int(value))
    elif code == -3:
        writer.write_int(int(value))
    elif code == -4:
        writer.write_uint(int(value))
    elif code == -5:
        writer.write_float(float(value))
    elif code == -6:
        writer.write_string(str(value))
    elif code == -7:
        writer.write_int64(int(value))
    elif code == -8:
        writer.write_uint64(int(value))
    else:
        raise ValueError(f"unknown builtin {code}")


def encode_definition(
    writer: ByteWriter, value: Any, def_idx: int, defs: list[Definition]
) -> None:
    d = defs[def_idx]
    if d.kind == KIND_ENUM:
        if isinstance(value, str):
            for f in d.fields:
                if f.name == value:
                    writer.write_varint(f.value)
                    return
            raise ValueError(f"unknown enum value: {value} for {d.name}")
        if isinstance(value, int):
            writer.write_varint(value)
            return
        raise ValueError(f"cannot encode {value!r} as enum {d.name}")
    if d.kind == KIND_STRUCT:
        for f in d.fields:
            encode_value(writer, value[f.name], f.type_code, f.is_array, defs)
        return
    if d.kind == KIND_MESSAGE:
        for f in d.fields:
            if f.name in value and value[f.name] is not None:
                writer.write_varint(f.value)
                encode_value(writer, value[f.name], f.type_code, f.is_array, defs)
        writer.write_varint(0)
        return
    raise ValueError(f"unknown kind {d.kind}")


def decode_definition(
    reader: ByteReader, def_idx: int, defs: list[Definition]
) -> Any:
    d = defs[def_idx]
    if d.kind == KIND_ENUM:
        value = reader.read_varint()
        for f in d.fields:
            if f.value == value:
                return f.name
        return value  # unknown enum value
    if d.kind == KIND_STRUCT:
        result: dict[str, Any] = {}
        for f in d.fields:
            result[f.name] = decode_value(reader, f.type_code, f.is_array, defs)
        return result
    if d.kind == KIND_MESSAGE:
        result = {}
        while True:
            tag = reader.read_varint()
            if tag == 0:
                return result
            target = None
            for f in d.fields:
                if f.value == tag:
                    target = f
                    break
            if target is None:
                raise ValueError(
                    f"unknown message field id {tag} in '{d.name}' (offset {reader.pos})"
                )
            result[target.name] = decode_value(
                reader, target.type_code, target.is_array, defs
            )
    raise ValueError(f"unknown kind {d.kind}")

File: packages/kiwi/src/codecs/kiwi.ts
// biome-ignore-all lint/suspicious/noBitwiseOperators: Kiwi binary encoding requires explicit bitwise operations.
import { KIND_ENUM, KIND_MESSAGE, KIND_STRUCT } from "../constants/kiwi";
import type { Definition, FieldDef, KiwiValue } from "../types/kiwi";

export type { Definition, FieldDef, KiwiValue } from "../types/kiwi";

const textDecoder = new TextDecoder("utf-8");
const textEncoder = new TextEncoder();
const scratch = new ArrayBuffer(4);
const scratchView = new DataView(scratch);

export class ByteReader {
  buf: Uint8Array;
  pos: number;

  constructor(buf: Uint8Array, pos = 0) {
    this.buf = buf;
    this.pos = pos;
  }

  remaining(): number {
    return this.buf.length - this.pos;
  }

  private at(pos: number): number {
    const b = this.buf[pos];
    if (b === undefined) {
      throw new Error(`buffer underrun at offset ${pos}`);
    }
    return b;
  }

  readByte(): number {
    const b = this.at(this.pos);
    this.pos += 1;
    return b;
  }

  readBytes(n: number): Uint8Array {
    const out = this.buf.subarray(this.pos, this.pos + n);
    this.pos += n;
    return out;
  }

  readVarint(): number {
    let result = 0;
    let shift = 0;
    while (true) {
      const b = this.at(this.pos);
      this.pos += 1;
      result |= (b & 0x7f) << shift;
      if ((b & 0x80) === 0) {
        return result >>> 0;
      }
      shift += 7;
      if (shift > 31) {
        throw new Error("varint too long");
      }
    }
  }

  readVarintBig(): bigint {
    let result = 0n;
    let shift = 0n;
    while (true) {
      const b = this.at(this.pos);
      this.pos += 1;
      result |= BigInt(b & 0x7f) << shift;
      if ((b & 0x80) === 0) {
        return result;
      }
      shift += 7n;
      if (shift > 63n) {
        throw new Error("varint too long");
      }
    }
  }

  readSignedVarint(): number {
    const v = this.readVarint();
    return (v >>> 1) ^ -(v & 1);
  }

  readBool(): boolean {
    return this.readByte() !== 0;
  }

  readInt(): number {
    return this.readSignedVarint();
  }

  readUint(): number {
    return this.readVarint();
  }

  readInt64(): bigint {
    const v = this.readVarintBig();
    return (v >> 1n) ^ -(v & 1n);
  }

  readUint64(): bigint {
    return this.readVarintBig();
  }

  readFloat(): number {
    const first = this.readByte();
    if (first === 0) {
      return 0;
    }
    const b1 = this.readByte();
    const b2 = this.readByte();
    const b3 = this.readByte();
    const raw = (first | (b1 << 8) | (b2 << 16) | (b3 << 24)) >>> 0;
    const bits = ((raw << 23) | (raw >>> 9)) >>> 0;
    scratchView.setUint32(0, bits, true);
    return scratchView.getFloat32(0, true);
  }

  readString(): string {
    const start = this.pos;
    let end = start;
    while (this.buf[end] !== 0) {
      end += 1;
    }
    const s = textDecoder.decode(this.buf.subarray(start, end));
    this.pos = end + 1;
    return s;
  }
}

export class ByteWriter {
  private readonly chunks: number[] = [];

  writeByte(b: number): void {
    this.chunks.push(b & 0xff);
  }

  writeBytes(bytes: Uint8Array): void {
    for (const b of bytes) {
      this.chunks.push(b);
    }
  }

  writeVarint(v: number): void {
    if (v < 0) {
      throw new Error("varint must be non-negative");
    }
    let value = v >>> 0;
    while (true) {
      const b = value & 0x7f;
      value >>>= 7;
      if (value) {
        this.chunks.push(b | 0x80);
      } else {
        this.chunks.push(b);
        return;
      }
    }
  }

  writeVarintBig(v: bigint): void {
    if (v < 0n) {
      throw new Error("varint must be non-negative");
    }
    let value = v;
    while (true) {
      const b = Number(value & 0x7fn);
      value >>= 7n;
      if (value > 0n) {
        this.chunks.push(b | 0x80);
      } else {
        this.chunks.push(b);
        return;
      }
    }
  }

  writeSignedVarint(v: number): void {
    const zig = v < 0 ? (-v << 1) - 1 : v << 1;
    this.writeVarint(zig >>> 0);
  }

  writeBool(v: boolean): void {
    this.writeByte(v ? 1 : 0);
  }

  writeInt(v: number): void {
    this.writeSignedVarint(v);
  }

  writeUint(v: number): void {
    this.writeVarint(v);
  }

  writeInt64(v: bigint): void {
    const zig = v < 0n ? (-v << 1n) - 1n : v << 1n;
    this.writeVarintBig(zig);
  }

  writeUint64(v: bigint): void {
    this.writeVarintBig(v);
  }

  writeFloat(v: number): void {
    if (v === 0) {
      this.writeByte(0);
      return;
    }
    scratchView.setFloat32(0, v, true);
    const bits = scratchView.getUint32(0, true);
    const rotated = ((bits << 9) | (bits >>> 23)) >>> 0;
    this.chunks.push(rotated & 0xff);
    this.chunks.push((rotated >>> 8) & 0xff);
    this.chunks.push((rotated >>> 16) & 0xff);
    this.chunks.push((rotated >>> 24) & 0xff);
  }

  writeString(s: string): void {
    const bytes = textEncoder.encode(s);
    for (const b of bytes) {
      this.chunks.push(b);
    }
    this.chunks.push(0);
  }

  get(): Uint8Array {
    return Uint8Array.from(this.chunks);
  }
}

export function parseSchema(buf: Uint8Array): Definition[] {
  const r = new ByteReader(buf);
  const count = r.readVarint();
  const defs: Definition[] = [];
  for (let i = 0; i < count; i += 1) {
    const name = r.readString();
    const kind = r.readByte();
    const fieldCount = r.readVarint();
    const fields: FieldDef[] = [];
    for (let j = 0; j < fieldCount; j += 1) {
      const fname = r.readString();
      const ftype = r.readSignedVarint();
      const farr = r.readByte();
      const fval = r.readVarint();
      fields.push({
        name: fname,
        typeCode: ftype,
        isArray: farr === 1,
        value: fval,
      });
    }
    defs.push({ name, kind, fields });
  }
  return defs;
}

function decodeBuiltin(reader: ByteReader, code: number): KiwiValue {
  switch (code) {
    case -1:
      return reader.readBool();
    case -2:
      return reader.readByte();
    case -3:
      return reader.readInt();
    case -4:
      return reader.readUint();
    case -5:
      return reader.readFloat();
    case -6:
      return reader.readString();
    case -7:
      return reader.readInt64();
    case -8:
      return reader.readUint64();
    default:
      throw new Error(`unknown builtin ${code}`);
  }
}

function encodeBuiltin(
  writer: ByteWriter,
  value: KiwiValue,
  code: number
): void {
  switch (code) {
    case -1:
      writer.writeBool(Boolean(value));
      return;
    case -2:
      writer.writeByte(numberValue(value, "byte"));
      return;
    case -3:
      writer.writeInt(numberValue(value, "int"));
      return;
    case -4:
      writer.writeUint(numberValue(value, "uint"));
      return;
    case -5:
      writer.writeFloat(numberValue(value, "float"));
      return;
    case -6:
      writer.writeString(stringValue(value));
      return;
    case -7:
      writer.writeInt64(bigintValue(value, "int64"));
      return;
    case -8:
      writer.writeUint64(bigintValue(value, "uint64"));
      return;
    default:
      throw new Error(`unknown builtin ${code}`);
  }
}

function numberValue(value: KiwiValue, typeName: string): number {
  if (typeof value !== "number") {
    throw new Error(`cannot encode ${String(value)} as ${typeName}`);
  }
  return value;
}

function stringValue(value: KiwiValue): string {
  if (typeof value !== "string") {
    throw new Error(`cannot encode ${String(value)} as string`);
  }
  return value;
}

function bigintValue(value: KiwiValue, typeName: string): bigint {
  if (typeof value === "bigint") {
    return value;
  }
  if (typeof value === "number") {
    return BigInt(value);
  }
  throw new Error(`cannot encode ${String(value)} as ${typeName}`);
}

function arrayValue(value: KiwiValue): KiwiValue[] {
  if (!Array.isArray(value)) {
    throw new Error(`cannot encode ${String(value)} as array`);
  }
  return value;
}

function objectValue(value: KiwiValue): Record<string, KiwiValue> {
  if (!isKiwiRecord(value)) {
    throw new Error(`cannot encode ${String(value)} as object`);
  }
  return value;
}

function isKiwiRecord(value: KiwiValue): value is Record<string, KiwiValue> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function decodeValue(
  reader: ByteReader,
  typeCode: number,
  isArray: boolean,
  defs: Definition[]
): KiwiValue {
  if (isArray) {
    const n = reader.readVarint();
    const out: KiwiValue[] = [];
    for (let i = 0; i < n; i += 1) {
      out.push(decodeValue(reader, typeCode, false, defs));
    }
    return out;
  }
  if (typeCode < 0) {
    return decodeBuiltin(reader, typeCode);
  }
  return decodeDefinition(reader, typeCode, defs);
}

export function encodeValue(
  writer: ByteWriter,
  value: KiwiValue,
  typeCode: number,
  isArray: boolean,
  defs: Definition[]
): void {
  if (isArray) {
    if (typeCode === -2 && value instanceof Uint8Array) {
      writer.writeVarint(value.length);
      writer.writeBytes(value);
      return;
    }
    const arr = arrayValue(value);
    writer.writeVarint(arr.length);
    for (const item of arr) {
      encodeValue(writer, item, typeCode, false, defs);
    }
    return;
  }
  if (typeCode < 0) {
    encodeBuiltin(writer, value, typeCode);
    return;
  }
  encodeDefinition(writer, value, typeCode, defs);
}

function defAt(defs: Definition[], idx: number): Definition {
  const d = defs[idx];
  if (!d) {
    throw new Error(`definition index out of range: ${idx}`);
  }
  return d;
}

export function decodeDefinition(
  reader: ByteReader,
  defIdx: number,
  defs: Definition[]
): KiwiValue {
  const d = defAt(defs, defIdx);
  if (d.kind === KIND_ENUM) {
    const value = reader.readVarint();
    for (const f of d.fields) {
      if (f.value === value) {
        return f.name;
      }
    }
    return value;
  }
  if (d.kind === KIND_STRUCT) {
    const result: Record<string, KiwiValue> = {};
    for (const f of d.fields) {
      result[f.name] = decodeValue(reader, f.typeCode, f.isArray, defs);
    }
    return result;
  }
  if (d.kind === KIND_MESSAGE) {
    const result: Record<string, KiwiValue> = {};
    while (true) {
      const tag = reader.readVarint();
      if (tag === 0) {
        return result;
      }
      let target: FieldDef | null = null;
      for (const f of d.fields) {
        if (f.value === tag) {
          target = f;
          break;
        }
      }
      if (!target) {
        throw new Error(
          `unknown message field id ${tag} in '${d.name}' (offset ${reader.pos})`
        );
      }
      result[target.name] = decodeValue(
        reader,
        target.typeCode,
        target.isArray,
        defs
      );
    }
  }
  throw new Error(`unknown kind ${d.kind}`);
}

export function encodeDefinition(
  writer: ByteWriter,
  value: KiwiValue,
  defIdx: number,
  defs: Definition[]
): void {
  const d = defAt(defs, defIdx);
  if (d.kind === KIND_ENUM) {
    if (typeof value === "string") {
      for (const f of d.fields) {
        if (f.name === value) {
          writer.writeVarint(f.value);
          return;
        }
      }
      throw new Error(`unknown enum value: ${value} for ${d.name}`);
    }
    if (typeof value === "number") {
      writer.writeVarint(value);
      return;
    }
    throw new Error(`cannot encode ${String(value)} as enum ${d.name}`);
  }
  if (d.kind === KIND_STRUCT) {
    const obj = objectValue(value);
    for (const f of d.fields) {
      encodeValue(writer, obj[f.name], f.typeCode, f.isArray, defs);
    }
    return;
  }
  if (d.kind === KIND_MESSAGE) {
    const obj = objectValue(value);
    for (const f of d.fields) {
      const v = obj[f.name];
      if (v !== undefined && v !== null) {
        writer.writeVarint(f.value);
        encodeValue(writer, v, f.typeCode, f.isArray, defs);
      }
    }
    writer.writeVarint(0);
    return;
  }
  throw new Error(`unknown kind ${d.kind}`);
}

export function findDefinition(defs: Definition[], name: string): number {
  for (let i = 0; i < defs.length; i += 1) {
    const d = defs[i];
    if (d && d.name === name) {
      return i;
    }
  }
  throw new Error(`definition not found: ${name}`);
}
